lib/maxmind/db/decoder.rb

```diff
@@ -0,0 +1,246 @@
+      # The specification states the formula in the `Data Section Separator'
+      # section.
+      #
+      # +pointer_test+ is used for testing pointer code.
+      def initialize(io, pointer_base=0, pointer_test=false)
+        @io = io
+        @pointer_base = pointer_base
+        @pointer_test = pointer_test
+      end
+
+      private
+
+      def decode_array(size, offset)
+        array = []
+        size.times do
+          value, offset = decode(offset)
+          array << value
+        end
+        return array, offset
+      end
+
+      def decode_boolean(size, offset)
+        return size != 0, offset
+      end
+
+      def decode_bytes(size, offset)
+        return @io.read(offset, size), offset + size
+      end
+
+      def decode_double(size, offset)
+        verify_size(8, size)
+        buf = @io.read(offset, 8)
+        return buf.unpack('G'.freeze)[0], offset + 8
+      end
+
+      def decode_float(size, offset)
+        verify_size(4, size)
+        buf = @io.read(offset, 4)
+        return buf.unpack('g'.freeze)[0], offset + 4
+      end
+
+      def verify_size(expected, actual)
+        if expected != actual
+          raise InvalidDatabaseError,
+            'The MaxMind DB file\'s data section contains bad data (unknown data type or corrupt data)'.freeze
+        end
+      end
+
+      def decode_int32(size, offset)
+        decode_int('l>'.freeze, 4, size, offset)
+      end
+
+      def decode_uint16(size, offset)
+        decode_int('n'.freeze, 2, size, offset)
+      end
+
+      def decode_uint32(size, offset)
+        decode_int('N'.freeze, 4, size, offset)
+      end
+
+      def decode_uint64(size, offset)
+        decode_int('Q>'.freeze, 8, size, offset)
+      end
+
+      def decode_int(type_code, type_size, size, offset)
+        if size == 0
+          return 0, offset
+        end
+
+        buf = @io.read(offset, size)
+        if size != type_size
+          buf = buf.rjust(type_size, "\x00".freeze)
+        end
+        return buf.unpack(type_code)[0], offset + size
+      end
+
+      def decode_uint128(size, offset)
+        if size == 0
+          return 0, offset
+        end
+
+        buf = @io.read(offset, size)
+
+        if size <= 8
+          buf = buf.rjust(8, "\x00".freeze)
+          return buf.unpack('Q>'.freeze)[0], offset + size
+        end
+
+        a_bytes = buf[0...-8].rjust(8, "\x00".freeze)
+        b_bytes = buf[-8...buf.length]
+        a = a_bytes.unpack('Q>'.freeze)[0]
+        b = b_bytes.unpack('Q>'.freeze)[0]
+        a <<= 64
+        return a | b, offset + size
+      end
+
+      def decode_map(size, offset)
+        container = {}
+        size.times do
+          key, offset = decode(offset)
+          value, offset = decode(offset)
+          container[key] = value
+        end
+        return container, offset
+      end
+
+      def decode_pointer(size, offset)
+        pointer_size = size >> 3
+
+        if pointer_size == 0
```

Comment:
I don't think it's necessary, but you could probably use a `case/when` here.

Reply:
May as well!